### ISSUE TITLE
to rely on audispd

### ISSUE THREAD

**Issue by mutazalsallal** [osquery/osquery#5594]
# Feature request

### What new feature do you want?

I was thinking if we can reconsider osquery approach with the linux audits, I think if we can rely on audispd, it will be helpful, so we will have the ability to run multiple agents, or utilize other things from linux native audits,
eg: if we want to detect strace to SSHD process, we need the linux audits at this stage as we dont have any tables for it, but we can't use the linux audits if we have osquery installed

### How is this new feature useful?

it will help help us to use osquery along side other agents

**Comment by mike-myers-tob** [comment]
### Problem
To use osquery's Audit-based evented tables, osquery can be the only consumer of Audit events while running. The primary problem this creates is that it precludes using SELinux.

### Goal
To use the audit dispatcher daemon `audispd` (from the `auditd` package) to collect Audit events simultaneously in (a) both osquery and other Audit consumers, or, (b) both core and an osquery extension. This would be a secondary, optional configuration for osquery because it would be dependent on the `auditd` package being installed.

### Considerations for Discussion

1. Event consumers that wish to receive events from [`audispd`](https://linux.die.net/man/8/audispd) must register a "plugin" with `audispd`, which `audispd` will launch and send events to. To preserve the way `osqueryd` is managed/started/stopped as a `systemd` service, would need to create or reuse an `audispd` plug-in process (external to `osqueryd`), that `osqueryd` would collect events from. Perhaps we'd want to use [the `audisp-remote` plugin](https://linux.die.net/man/8/audisp-remote) from the `audispd-plugins` package, and allow `osqueryd` to work as an "aggregating server" over a localhost socket.

2. A concern, still unconfirmed (should be possible to resolve this with testing): does the multiplexing performed by `audispd` deliver the same events to all event listeners? If it does, then each listener would receive the superset of events that all listeners are subscribed to. Is this a performance problem? If so, how much of a problem is it?

3. Syscall filtering and configuring Audit: currently, osquery directly configures the Audit subsystem to turn on specific events for listening. When using `audispd`, configuration of Audit must happen via `auditd`, which then passes the data to `audisp`. Configuration via `auditd` is performed using the external process `auditctl` — would the osquery project maintainers agree to loosen the design principles of osquery ("no spawning of additional processes") to allow it to call this process? Or require that the user manually configure things outside of osquery?

Tagging @alessandrogario 

Also tagging @uptycs-nishant because we have heard secondhand that Uptycs has experimented with using osquery with `audispd`, and may have insights here.

**Comment by directionless** [comment]
Thinking aloud... I'm not super familiar with `audispd`, but from what I understand I think there are two reasonable ways for us to integrate.

1. Use `audisp-remote` to send events into a special osquery interface. I don't love needing to open a port, and I don't relish authentication
2. Use a shim that `audisp` and can talk over osquery's extension socket.

I think I favor (2) here. But I don't think that's my call to me.

**Comment by uptycs-nishant** [comment]
> A concern, still unconfirmed (should be possible to resolve this with testing): does the multiplexing 
> performed by audispd deliver the same events to all event listeners? If it does, then each listener 
> would receive the superset of events that all listeners are subscribed to. 

yes, audispd deliver the same events to all event listeners. yes, each listener would receive the superset of events that all listeners are subscribed to.

There are two buitin plugins provided by audispd -
1. syslog plugin 
  If syslog plugin is activated then audispd will forward audit records to syslogs as well
2. unix domain socket 
  If this plugin is activated then an external program like osqueryd can read the audit records from unix domain socket.
Uptycs uses the following path to receive audit records via audispd.
kauditd (God land) ---{over netlink socket}--->auditd(Mortal land)---{over stdin}-->audispd(Multiplexer)--{over unix domain socket}----->osqueryd


> Is this a performance problem? If so, how much of a problem is it?
At present audispd is a process so messages will be crossing address space boudaries in newer version of auditd package audispd is going to be a thread so same address space therefore little more performant.

This is way less efficient 

kauditd ---{over netlink socket}--->osqueryd
VS
kauditd ---{over netlink socket}--->auditd---{over stdin}-->audispd(Multiplexer)--{over unix domain socket}----->osqueryd

If you choose not to use unix domain socket path and have your own exe which will forward the events to osqueryd after receiving from audispd then messages path look like below -

kauditd ---{over netlink socket}--->auditd---{over stdin}-->audispd(Multiplexer)--over stdin--><event forwarding exe>--{some ipc mechanism, preferably shared memory}--->osqueryd

**Comment by zwass** [comment]
Thanks for the helpful info and analysis folks!

FWIW I'm taking a look at the [source code for the unix domain socket plugin](https://github.com/linux-audit/audit-userspace/blob/master/audisp/plugins/af_unix/audisp-af_unix.c) and it looks like it runs as a separate process just like any other audispd plugin (reading from stdin and writing to the socket if there's a listener). So there's probably one more step  in the diagrams from above.

**Comment by directionless** [comment]
Thank you for this thoughtful suggestion about using audispd for audit event consumption. The coexistence problem you identified — osquery monopolizing the audit socket and preventing other agents from receiving events — has been a known pain point.

However, this issue has been open for over 6 years without progress toward implementation. osquery's audit subsystem handling has evolved since this was filed, and the landscape of Linux auditing tools has changed as well.

I'm going to close this for now due to inactivity. If coexistence with other audit consumers is still a concern you're experiencing with current osquery versions, please feel free to open a new issue with details about your current setup and osquery version. osquery is a volunteer project, and contributions toward improving audit multiplexing would be very welcome!
